flink-ml-lib/src/main/java/org/apache/flink/ml/feature/minmaxscaler/MinMaxScaler.java

```diff
@@ -0,0 +1,165 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.feature.minmaxscaler;
+
+import org.apache.flink.api.common.functions.RichMapPartitionFunction;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.ml.api.Estimator;
+import org.apache.flink.ml.common.datastream.DataStreamUtils;
+import org.apache.flink.ml.linalg.DenseVector;
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.ml.util.ReadWriteUtils;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/** An Estimator which implements the MinMaxScaler algorithm. */
+public class MinMaxScaler
+        implements Estimator<MinMaxScaler, MinMaxScalerModel>, MinMaxScalerParams<MinMaxScaler> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public MinMaxScaler() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public MinMaxScalerModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Tuple2<DenseVector, DenseVector>> minMaxVectors =
+                computeMinMaxVectors(tEnv.toDataStream(inputs[0]), getFeaturesCol());
+        DataStream<MinMaxScalerModelData> modelData = genModelData(minMaxVectors);
+        MinMaxScalerModel model =
+                new MinMaxScalerModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static MinMaxScaler load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Generates minMax scaler model data.
+     *
+     * @param minMaxVectors Input distributed minMaxVectors.
+     * @return MinMax scaler model data.
+     */
+    private static DataStream<MinMaxScalerModelData> genModelData(
+            DataStream<Tuple2<DenseVector, DenseVector>> minMaxVectors) {
+        DataStream<MinMaxScalerModelData> modelData =
+                DataStreamUtils.mapPartition(
```

Comment:
nits: using `mapPartition` here may be not that efficient since we cache all the input records in the state.

How about using a self-defined streamOperator as [1]?

[1] https://github.com/zhipeng93/flink-ml/blob/ccd3f9919dbdd3c5123f1cc88636f832f8215ffb/flink-ml-core/src/main/java/org/apache/flink/ml/common/datastream/DataStreamUtils.java#L68

Reply:
This mapPartition only has fewer record. For every worker only has one record. I think it's no efficient problem.